### ISSUE TITLE
rm(..., recursive=true) "directory not empty" errors

### ISSUE THREAD

**Issue by stevengj** [JuliaLang/julia#34700]
I was trying to delete my Anaconda installation, which is a huge directory tree, with
```jl
using Conda
rm(Conda.ROOTENV, recursive=true)
```
and I keep getting `SystemError .... rmdir: directory not empty` errors.     However, if I check that directory, I find that it is indeed empty.   Indeed, if I run `rm(Conda.ROOTENV, recursive=true)` *again*, then it succeeds in deleting that directory, but then proceeds to give me a similar error with another directory, ad nauseam.

  For example (apologies for the screenshot):
![image](https://user-images.githubusercontent.com/2913679/74077823-31d1cb00-49f1-11ea-83ac-338e186afa6b.png)

It looks like there is some kind of race condition in the recursive `rm`, where it empties the directory, but Windows doesn't realize that the files are gone when it tries to `rmdir`.

**Comment by DilumAluthge** [comment]
Out of curiosity, what happens if you try running `rm` with the keyword argument `force = true`?

**Comment by DilumAluthge** [comment]
Also, can you post the output of `] st -m`?

**Comment by stevengj** [comment]
I tried it with `force=true` (actually I did this in the screenshot), and there was no difference.

**Comment by stevengj** [comment]
`st -m` output is here, though that shouldn't matter much as `rm` is a `Base` function.  (Again, apologies for the screenshot — I'm running Windows in a VM and having trouble getting copy-and-paste to work.   But a student's native Windows 10 laptop with a fresh Julia installation exhibited the same problem yesterday.)

![image](https://user-images.githubusercontent.com/2913679/74086346-fdd9c280-4a4f-11ea-88b9-c45d2a73597b.png)

**Comment by EdoAlvarezR** [comment]
Running into the same issue

**Comment by EdoAlvarezR** [comment]
Turns out that `rm(...)` works just fine. It was a system-level issue: there was a `.fuse_hidden` file in the directory I wanted to remove that was keeping the system from being able to remove it (even the terminal `rm` was throwing the same error).

**Comment by stevengj** [comment]
> Turns out that rm(...) works just fine. It was a system-level issue

I don't think that was the case in my original issue, where simply running the `rm` command twice worked.

**Comment by AbhimanyuAryan** [comment]
running into same issue `ERROR: LoadError: SystemError (with ../build): rmdir: Directory not empty` with `rm("../build", force=true)` the directory contains sub directories with `.html` files

**Comment by vtjnash** [comment]
> where it empties the directory, but Windows doesn't realize that the files are gone when it tries to rmdir.

Windows does not delete files when you `rm` them, though it is possible to implement that as a feature by abusing the `rename` syscall. For example: https://github.com/vtjnash/Pidfile.jl/commit/81ef0a5e729e906d24c572d8408107d37e231d85#diff-3b86733f3bbb623c9149814588ef59698913a5e9d9654f999cc726c50a8334cdR241

**Comment by KristofferC** [comment]
That talks about open files. I don't think that is the case in the other scenarios here.

**Comment by vtjnash** [comment]
If you are running a virus scanner, many of those causes deletion errors without the trick above

**Comment by AhmedSalih3d** [comment]
I am seeing this error again in Julia 1.8.

**Comment by stevengj** [comment]
Can we call a Win32 function on Windows and let the OS deal with it?   [`SHFileOperationW`](https://learn.microsoft.com/en-us/windows/win32/api/shellapi/nf-shellapi-shfileoperationw) is supposed to be able to recursively delete directories.

**Comment by vtjnash** [comment]
I am uncertain if we load shell32.dll (I think libuv does), but it should just be delegating to our same underlying Win32 calls

**Comment by giordano** [comment]
For what is worth, I recently had this problem outside of Julia.  On an x86_64 linux system I got
```console
$ rm -rf directory/
rm: cannot remove 'directory/.spack-env': Directory not empty
```
What I believe has happened in my case is that a rogue `spack` process (which I thought I had killed but I failed to do so) was keeping writing to the `directory/.spack-env` directory while I was trying to delete it, thus causing the surprising error message.

What I want to say here is that if anything Julia isn't any worth than Coreutils `rm`.

**Comment by stevengj** [comment]
@giordano, we aren't talking about rogue processes here.   Even if the directory is completely static, Julia's `rm` is sometimes failing on Windows with large directory trees, because Windows apparently takes some time to realize that a directory is empty when you tell it to delete files.

**Comment by mestinso** [comment]
I've been facing this issue recently in a corporate windows environment. I'm confident it's antivirus/security related. I've experienced it in two places:
1) PythonCall environment cleanup at the end of running runtests.jl in a package that uses PythonCall
2) rm operation that is issued as part of a cleanup process when using when using FMI.jl toolbox

Not sure if there is anything I can do to help/expedite resolution/workaround, but I'd certainly offer to test any potential solutions.

**Comment by stevengj** [comment]
@mestinso, I don't think this has anything to do with #50842.  The issue here is that a *single* process has trouble recursively deleting a directory on Windows, apparently because of some quirks in the Windows filesystem.

I still think we should just call [`SHFileOperationW`](https://learn.microsoft.com/en-us/windows/win32/api/shellapi/nf-shellapi-shfileoperationw), since presumably Microsoft has figured out how to delete directories on their own operating system.

**Comment by p-foresman** [comment]
Just a heads up, I'm seeing this issue running Julia 1.10 on Linux CentOs 7.

**Comment by stevengj** [comment]
@p-foresman, without more information, I’m skeptical it’s the same issue. (There are legitimate reasons why you might get “directory not empty errors.”)

**Comment by KristofferC** [comment]
Not sure this is related but I got this on Windows

```
julia> rm(".julia/compiled/v1.11"; recursive=true)
ERROR: IOError: rm(".julia/compiled/v1.11\\Accessors"): directory not empty (ENOTEMPTY)
Stacktrace:
 [1] uv_error
   @ Base .\libuv.jl:100 [inlined]
 [2] rm(path::String; force::Bool, recursive::Bool)
   @ Base.Filesystem .\file.jl:307
```

when I had another Windows Julia session opened that kept the shared library in that folder open in the process.

**Comment by StefanKarpinski** [comment]
Yeah, unfortunately, Windows just won't let you delete something that's opened by some other process.

**Comment by stevengj** [comment]
See also this [stackoverflow thread](https://stackoverflow.com/questions/76864977/how-can-i-delete-a-file-that-is-currently-in-use-by-another-process-in-cpp-win32) — you can apparently mark the file to be deleted on the next reboot 😆 .

**Comment by StefanKarpinski** [comment]
Amazing 😂😭

**Comment by vtjnash** [comment]
Often you can just rename it too, as the actual lock preventing deletion is often on the name, not the file handle

**Comment by tlnagy** [comment]
I'm wondering if this is related to `EBUSY` errors when trying to run

```julia
rm(absolute_root; force=true, recursive=true) 
```

on Windows (ref: https://github.com/JuliaDocs/DemoCards.jl/issues/160)

**Comment by stevengj** [comment]
@tlnagy if calling `rm(absolute_root; force=true, recursive=true) ` repeatedly does not resolve your problem, then it's unrelated.

The issue here is recursive deletion of directories where nothing is locked/busy, but it still fails because of a race condition in the Windows filesystem.

**Comment by vtjnash** [comment]
FWIW, the race condition usually is not in the Windows filesystem, but rather that the NT kernel currently makes it impossible to implement a reliable virus scanner, but everyone runs a virus scanner anyways.

**Comment by inkydragon** [comment]
Same issue in Rust: https://github.com/rust-lang/rust/issues/29497
May learn something from: https://github.com/XAMPPRocky/remove_dir_all
> For Windows an implementation that handles the locking of directories that occurs when deleting directory trees rapidly.

**Comment by vtjnash** [comment]
That functionality hasn't been present in `remove_dir_all` since it was deleted in https://github.com/XAMPPRocky/remove_dir_all/commit/61c03eb9bf9591d1a143e2f3079290b2c4d2fe0e#diff-a4c907d91078072b617381605d6ded19c006a1934007d2607c73ebde5f7fd6bcL39-L41 (see the modified comments in that commit), but the documentation and README wasn't updated

But the rust stdlib now uses the POSIX_SEMANTICS flag, when supported by the OS and filesystem, to probably fix this: https://github.com/rust-lang/rust/commit/5ab67bff1e2230217ce133165605f8dc0d588178#diff-e8df55f38a9a224cf1cfd40e6c535535aa66e8073cc8d9b959308659ba1de1f9R564-R592

Seems like at least half of this bug might be in the implementation of DeleteFile on Win32 being bad, so that calling the underlying API instead is a bit more reliable? (DeleteFile sets the FILE_SHARE_DELETE flag, which delays deletion, while the underlying API can be called without that flag)

FWIW, this has all been known for at least a few years, but waiting for some developer to care enough about Windows to use the new kernel APIs: https://github.com/libuv/libuv/issues/3839

**Comment by vtjnash** [comment]
The Rust implementation of that change was here https://github.com/rust-lang/rust/commit/5ab67bff1e2230217ce133165605f8dc0d588178#diff-e8df55f38a9a224cf1cfd40e6c535535aa66e8073cc8d9b959308659ba1de1f9R572-R577

**Comment by kenjirotochimoto** [comment]
一時フォルダを変えてみたところ、エラーの内容が変わりました。
```
        # dir_intermediate = mktempdir(cleanup=false)
        dir_intermediate = mkdir("tmp_tra")
```
以下が別のエラーです。
```
PS C:\tmp2\TodaBlastRegionGenerator.jl> julia -e 'using TodaBlastRegionGenerator; TodaBlastRegionGenerator.main(ARGS)' volumes_train --face_pattern test/input_samples/face_pattern_003.csv --id 1 --pts "test/input_samples/点群/No.1219(24423.25)/#Conv_uxmData20211130_1544_right-handed.pts" --fdd "test/input_samples/削孔実績データ/No.1219(24423.25)/fdd_003_202111301233.csv" --out /home/horikawa/Desktop/volumes_train --tunnel_line test/input_samples/tunnel_line_003.csv --fid test/input_samples/削孔実績データ/fid_003.csv --face_no 1219f}")
[ Info: Save design sphered argument
[ Info: Save face sphereio.TextIOWrapper name='<stdout>' mode='w' encoding='utf-8'>
[ Info: Save design sphered argument
[ Info: Save face sphere
[ Info: Save blastplan
[ Info: cleanup falseeeeeeeeeeeeeeeeeeee
[ Info: tmp_tra
[ Info: Generate design AB
[ Info: Generate design C
[ Info: Generate design D
[ Info: Generate design E
[ Info: Generate mesh before blast
[ Info: Generate points after blast
[ Info: Generate mesh after blast
[ Info: Generate cylinders
[ Info: Generate ply files with Blender
[ Info: Calculate volumes
ERROR: IOError: rm("tmp_tra\\json"): directory not empty (ENOTEMPTY)
Stacktrace:
 [1] uv_error
   @ Base .\libuv.jl:100 [inlined]
 [2] rm(path::String; force::Bool, recursive::Bool)
   @ Base.Filesystem .\file.jl:307
 [3] rm(path::String; force::Bool, recursive::Bool)
   @ Base.Filesystem .\file.jl:294
 [4] rm
   @ Base.Filesystem .\file.jl:273 [inlined]
 [5] main(args::Vector{String})
   @ TodaBlastRegionGenerator C:\tmp2\TodaBlastRegionGenerator.jl\src\TodaBlastRegionGenerator.jl:390
 [6] top-level scope
   @ none:1
```

**Comment by arlowhite** [comment]
I was encountering this "directory not empty" error with pkg `instantiate` in Julia 1.10
when trying to `rm("C:\\Users\\USER\\.julia\\artifacts\\jl_vxDd76\\bin")`
I do have Cylance running (antivirus), possibly a factor?

Switching to Julia 1.11 fixed it! currently 1.11.0-alpha2+0.x64.w64.mingw32

possibly fixed by https://github.com/JuliaLang/julia/pull/53456
